File: pymenuset/menu.py
from pymenuset import config
from pymenuset.item import *

import re


class Menu(object):
	def __init__(self, input_text=None):
		self.title = ''
		self.items = []

		self.__input_text = ''
		self.__output_text = ''

		self.parse(input_text)

	def parse(self, input_text=None):
		if type(input_text) is not None:
			self.__input_text = input_text

		items = []
		icons = {}

		for line in self.__input_text.split("\n"):
			search = line[:4]

			if search == 'titl':
				self.__parse_title(line)
			elif search == 'icon':
				self.__parse_icon(line, icons)
			elif search == 'item':
				self.__parse_item(line, items)

		# connect icons with actions
		for id in icons.keys():
			try:
				items[int(id)].icon = icons[id]
			except AttributeError:
				pass

		# properly nest submenus
		stack = [self]

		for item in items:
			if type(item) is Submenu:
				stack[-1].add(item)
				stack.append(item)
			elif type(item) is SubmenuEnd:
				del stack[-1]
			else:
				stack[-1].add(item)

		return self

	def __parse_title(self, line):
		match = re.match('title=(.*)', line)
		self.title = match.group(1)

	def __parse_icon(self, line, icons):
		match = re.match('icon_([0-9]+)=(.*)', line)
		icons[match.group(1)] = match.group(2)

	def __parse_item(self, line, items):
		match = re.match('item_([0-9]+)=([^\s]+)(.*)', line)

		id = int(match.group(1))
		action = match.group(2)

		try:
			match.group(3)
		except AttributeError:
			name = None
		else:
			name = match.group(3)

		try:
			items[id]
		except IndexError:
			items.extend([None] * (id - len(items) + 1))

		item_lookup = {
			'-1': Separator,
			'-2': Submenu,
			'-3': SubmenuEnd,
			'-4': Label,
		}

		try:
			items[id] = item_lookup[action](id, name, action)
		except KeyError:
			items[id] = Action(id, name, action)

	def style(self):
		for item in self.items:
			item.style()

		return self

	def strip_labels(self):
		return self.strip_matching('label')

	def strip_matching(self, item_type=''):
		self.items = {loc: menu for loc, menu in self.items.items() if item_type.lower() != loc.__class__.__name__.lower()}

		return self

	def add(self, item):
		self.items.append(item)

	def flatten(self):
		self.__output_text = 'title=' + self.title + "\n"
		count = -1

		for item in self.items:
			count += 1
			count, more = item.flatten(count)
			self.__output_text += more

		self.__output_text += "\n"

		return self.__output_text


class ToolbarMenu(Menu):
	pass


class ContextMenu(Menu):
	pass

File: models/Menu.js
// import

const _ = require('lodash');
const fs = require('fs-extra');
const async = require('async');

const Separator = require('./Separator');
const Submenu = require('./Submenu');
const SubmenuEnd = require('./SubmenuEnd');
const Label = require('./Label');
const Action = require('./Action');

// fns

function parseTitle(menu, str) {
  menu.title = /title=(.*)/i.exec(str).pop();
}

function parseIcon(menu, str, icons) {
  let [, index, icon] = /icon_([0-9]+)=(.*)/i.exec(str);
  icons[index] = icon;
}

function parseItem(menu, str, items) {
  let [, index, action, name] = /item_([0-9]+)=([^\s]+)(.*)/i.exec(str);

  index = parseInt(index);
  let length = items.length;

  if (length <= index) {
    for (length; length <= index; length++) {
      items.push(undefined);
    }
  }

  switch(action) {
    case '-1':
      return items.push(new Separator());
    case '-2':
      return items.push(new Submenu(name));
    case '-3':
      return items.push(new SubmenuEnd());
    case '-4':
      return items.push(new Label(name));
    default:
      return items.push(new Action(name, action));
  }
}

// export

module.exports = class Menu {
  constructor(inputText) {
    this.inputText = inputText;
    this.outputText = '';

    this.title = '';
    this.items = [];

    this.parse();
  }

  parse(inputText) {
    if (inputText) {
      this.inputText = inputText;
    }

    let icons = {};
    let items = [];

    this.inputText.split('\n').forEach((str) => {
      switch (str.slice(0, 4)) {
        case 'titl':
          return parseTitle(this, str);
        case 'icon':
          return parseIcon(this, str, icons);
        case 'item':
          return parseItem(this, str, items);
      }
    });

    _.forEach(icons, (icon, i) => {
      if (items[i]) {
        items[i].icon = icon;
      }
    });

    let stack = [this];

    _.compact(items).forEach((item) => {
      switch (item.constructor.name) {
        case 'Submenu':
          if (!_.last(stack)) {
            console.log(item);
          }
          _.last(stack).items.push(item);
          stack.push(item);
          break;

        case 'SubmenuEnd':
          stack.pop();
          break;

        default:
          _.last(stack).items.push(item);
          break;
      }
    });

    return this;
  }

  style() {
    this.items.forEach((item) => {
      item.style();
    });

    return this;
  }

  flatten() {
    this.outputText = this.title ? `title=${this.title}\n` : '';
    let counter = {
      count: 0
    };

    this.items.forEach((item) => {
      this.outputText += item.flatten(counter);
      counter.count++;
    });

    return this.outputText;
  }
};